### DISCUSSION TITLE
[AssetMapper] vendor packages, like jsrouting-bundle

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#52249] [Q&A]
I'm trying to replace yarn/webpack in an application, and am wondering how to add [FosRoutingBundle](https://github.com/FriendsOfSymfony/FOSJsRoutingBundle/blob/master/Resources/doc/installation.rst)

I assume importmap needs to follow this pattern

```php
 '@symfony/stimulus-bundle' => [
        'path' => './vendor/symfony/stimulus-bundle/assets/dist/loader.js',
    ],
```

The yarn installation is

    yarn add -D ./vendor/friendsofsymfony/jsrouting-bundle/Resources/

But what's the AssetMapper 6.4 way of doing the same?

I'm trying to clean up this section of the code in a stimulus controller.

```js
routes = require('../../public/js/fos_js_routes.json');
import Routing from '../../vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/router.min.js';
```

Is there an offficial package on jsdelivr?

**Comment by weaverryan** [comment]
Hey! 

I think this bundle will need some work for AssetMapper, though I don't use it personally (I typically have URLs on my HTML elements then read those attributes from JavaScript. Or I pass the URLs as values to my Stimulus controller).

The bundle will need to add an asset mapper path that points to their `Resources/public/` directory. But you can also add an asset mapper path in your own `asset_mapper.yaml` config file temporary to that directory. 

Once you do, yea, I'm pretty sure this will work:

```js
import Routing from '../../vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/router.min.js';
```

However, this will NOT work:

```js
// even if you change to import - you can't import JSON directly from a true module environment
routes = require('../../public/js/fos_js_routes.json');
```

Instead, the bundle should make a `.js` module available that exports the route information - e.g.

```js
export default {
    // the JSON route info
}
```

To make life even easier (and honestly, easier than with Encore), the bundle could create an asset compiler that generates that file dynamically at run-time. StimulusBundle does this - https://github.com/symfony/ux/blob/2.x/src/StimulusBundle/src/AssetMapper/StimulusLoaderJavaScriptCompiler.php#L37-L42 - transforming a `controller.js` file which is [boring and empty](https://github.com/symfony/ux/blob/2.x/src/StimulusBundle/assets/dist/controllers.js) into a file that dynamically returns the [correct imported controllers](https://ux.symfony.com/assets/@symfony/stimulus-bundle/controllers-5780b65594f242953a22957517e43d2e.js) - e.g.

```js
import controller_0 from '../ux-toggle-password/controller.js';
import controller_3 from '../../controllers/bootstrap-modal-controller.js';
export const eagerControllers = {"symfony--ux-toggle-password--toggle-password": controller_0,  "bootstrap-modal": controller_3 };
export const lazyControllers = {"symfony--ux-autocomplete--autocomplete": () => import("../ux-autocomplete/controller.js")};
export const isApplicationDebug = false;
```

I'd be happy to help anyone with this that wants to try it. The end result, I think, will be quite a nice DX.

**Reply by tacman** [reply]
I've added the path to asset.yaml
```yaml
framework:
    asset_mapper:
        # The paths to make available to the asset mapper.
        paths:
            - assets/
            - vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/
```

But I'm confused as to the relative path when importing:
```
       import Routing from 'vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/router.min.js';
       import Routing from '../../vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/router.min.js'
etc..
```

In this particular case, it's being called from a template in templates/Edition/browse.html.twig, so should the import be relative to this file? Once this works, I'll move it to a stimulus controller in a bundle.

**Comment by tacman** [comment]
Hmm.  Beyond me right now.  

I'm guessing that the js-translation-bundle (https://packagist.org/packages/willdurand/js-translation-bundle)  will have the same need, though maybe less important now that https://symfony.com/bundles/ux-translator/current/index.html exists

**Comment by tacman** [comment]
Brainstorming: What about refactoring the dump of the json routes to be a .js file with export const ROUTE_NAME = ..., following the model of ux-translation.  That is, instead of loading .json data, load a bunch of constants.

As a bonus, all the routes tagged with { export: true } will be available as constants, so path(USER_PROFILE, { id: user.id }) 

I can move this discussion to that bundle if this approach makes sense.

**Reply by stof** [reply]
This should definitely be discussed in the bundle rather than here.
